core/src/smartcontracts/isi/domain.rs

```diff
@@ -391,32 +391,34 @@ pub mod isi {
 /// Query module provides [`Query`] Domain related implementations.
 pub mod query {
     use eyre::Result;
-    use iroha_data_model::{domain::Domain, query::error::QueryExecutionFail as Error};
+    use iroha_data_model::{
+        domain::Domain,
+        query::{
+            error::QueryExecutionFail as Error,
+            predicate::{predicate_atoms::domain::DomainPredicateBox, CompoundPredicate},
+        },
+    };
     use iroha_primitives::json::JsonString;
 
     use super::*;
-    use crate::state::StateReadOnly;
+    use crate::{smartcontracts::ValidIterableQuery, state::StateReadOnly};
 
-    impl ValidQuery for FindAllDomains {
+    impl ValidIterableQuery for FindAllDomains {
         #[metrics(+"find_all_domains")]
         fn execute<'state>(
-            &self,
+            self,
+            filter: CompoundPredicate<DomainPredicateBox>,
             state_ro: &'state impl StateReadOnly,
-        ) -> Result<Box<dyn Iterator<Item = Domain> + 'state>, Error> {
-            Ok(Box::new(state_ro.world().domains_iter().cloned()))
-        }
-    }
-
-    impl ValidQuery for FindDomainById {
-        #[metrics(+"find_domain_by_id")]
-        fn execute(&self, state_ro: &impl StateReadOnly) -> Result<Domain, Error> {
-            let id = &self.id;
-            iroha_logger::trace!(%id);
-            Ok(state_ro.world().domain(id)?.clone())
+        ) -> std::result::Result<impl Iterator<Item = Domain> + 'state, Error> {
+            Ok(state_ro
```

Comment:
I've checked and it look like that some simple cases indeed could be optimized quite easily!
This can be left for follow up PR.

```suggestion
            let iter: Box<dyn Iterator<Item = Domain>> = match filter {
                CompoundPredicate::Atom(DomainPredicateBox::Id(DomainIdPredicateBox::Equals(
                    domain_id,
                ))) => Box::new(Some(state_ro
                    .world()
                    .domain(&domain_id)?
                    .clone())
                    .into_iter()),
                filter => Box::new(
                    state_ro
                        .world()
                        .domains_iter()
                        .filter(move |&v| filter.applies(v))
                        .cloned(),
                ),
            };

            Ok(iter)
```

Reply:
I've also envisioned a helper functions for this matching, probably this can make it a bit less boilerplaty

Reply:
should we open a new issue for this?